### ISSUE TITLE
"evaluate -> click -> wait -> evaulate" loop - is it possible?

### ISSUE THREAD

**Issue by kamiltrebunia** [segmentio/nightmare#837]
Hello! I'm trying to goto page, wait for selector, then evaluate to get number of steps I will have to do (walk through pagination which is dynamically loaded with xhr by target website) and then run evaluate->click->wait->repeat as many times as required. Is it at all possible?

It certainly doesn't seem possible since there's no way to get access to the result of evaluate function other than by calling 'end' followed by 'then'.

It seems to be a common use case (parsing paginated websites) - easily achievable with other tools (sadly all of them based on Phantom), so I suspect I may be missing something.

Any advise?

**Comment by rosshinkley** [comment]
First and foremost: you don't have to call `.end()` before `.then()`.  Use `.end()` when you want to shut the Electron instance down.  Use `.then()` when you want to perform the actions that have been queued.

Setting that aside, getting results from an asynchronously loaded page seems to come up fairly frequently.  I'd suggest first giving ["Asynchronous Operations and Loops" in `nightmare-examples`](https://github.com/rosshinkley/nightmare-examples/blob/master/docs/common-pitfalls/async-operations-loops.md) a read if you haven't already.  That should get you started on looping.  I'd also recommend reading #625.  It's a different issue (scraping an infinite scroll), but the bones of the issue are similar:  conduct an action, perform some sort of check, iterate if necessary.

Finally, if you have specific source you'd like a hand with, feel free to include a simplified version.  I'd be happy to help you get started.

**Comment by kamiltrebunia** [comment]
Thank you very much @rosshinkley. I didn't stumble upon nightmare-examples before. It's true that it's linked within a ToC, but by scrolling and reading through segmentio/nightmare README there was no such section, not even a link to it and I didn't pay attention to the ToC since I assumed I read it all. Also, there is not a single example of a code in a main README that would use `then` at least twice. If either of these were the case - I would have probably solved it by myself before. I'm just describing what was the issue in my case, feel free to disregard my remarks :)

_Even then I am not sure what was I doing wrong before since I used map.reduce (or recursion) to chain promises before and it eludes me what was I previously doing wrong in my code that I did draw all those wrong conclusions about the function of `then`._

In any case - it works like a charm and now that I finally get nightmare mechanics I enjoy the simplicity of the API very much. Fantastic tool.
